plugins/statistics.koplugin/main.lua

```diff
@@ -1862,38 +1893,71 @@ function ReaderStatistics:onPageUpdate(pageno)
+        return
+    end
+
+    self.pageturn_count = self.pageturn_count + 1
+    local now_ts = TimeVal:now().sec
+
+    -- Get the previous page's last timestamp (if there is one)
+    local page_data = self.page_stat[self.curr_page]
+    -- This is a list of tuples, in insertion order, we want the last one
+    local data_tuple = page_data and page_data[#page_data]
+    -- Tuple layout is { timestamp, duration }
+    local then_ts = data_tuple and data_tuple[1]
+    -- If we don't have a previous timestamp to compare to, abort early
+    if not then_ts then
+        logger.dbg("ReaderStatistics: No timestamp for previous page", self.curr_page)
+        self.page_stat[pageno] = { { now_ts, 0 } }
+        self.curr_page = pageno
+        return
+    end
+
+    -- By now, we're sure that we actually have a tuple (and the rest of the code ensures they're sane, i.e., zero-initialized)
+    local curr_duration = data_tuple[2]
+    -- NOTE: If all goes well, given the earlier curr_page != pageno check, curr_duration should always be 0 here.
+    -- Compute the difference between now and the previous page's last timestamp
+    local diff_time = now_ts - then_ts
+    if diff_time >= self.page_min_read_sec and diff_time <= self.page_max_read_sec then
+        self.mem_read_time = self.mem_read_time + diff_time
+        -- If it's the first time we're computing a duration for this page, count it as read
+        if #page_data == 1 and curr_duration == 0 then
+            self.mem_read_pages = self.mem_read_pages + 1
         end
-        table.sort(sorted_performance)
-        local read_pages_set = {}
-        local diff_time
-        for i=1, #sorted_performance - 1 do
-            diff_time = sorted_performance[i + 1] - sorted_performance[i]
-            if diff_time <= self.page_max_read_sec and diff_time >= self.page_min_read_sec  then
-                mem_read_time = mem_read_time + diff_time
-                read_pages_set[self.pages_stats[sorted_performance[i]]] = true
-            elseif diff_time > self.page_max_read_sec then
-                mem_read_time = mem_read_time + self.page_max_read_sec
-                read_pages_set[self.pages_stats[sorted_performance[i]]] = true
-            end
+        -- Update the tuple with the computed duration
+        data_tuple[2] = curr_duration + diff_time
+    elseif diff_time > self.page_max_read_sec then
+        self.mem_read_time = self.mem_read_time + self.page_max_read_sec
+        if #page_data == 1 and curr_duration == 0 then
+            self.mem_read_pages = self.mem_read_pages + 1
         end
-        mem_read_pages = util.tableSize(read_pages_set)
+        -- Update the tuple with the computed duration
+        data_tuple[2] = curr_duration + self.page_max_read_sec
     end
-    -- every 50 pages we write stats to database
-    if util.tableSize(self.pages_stats) % PAGE_INSERT == 0 then
+
+    -- We want a flush to db every 50 page turns
+    if self.pageturn_count >= PAGE_INSERT then
```

Comment:
(might give a more explicite name to that PAGE_INSERT variable, while you're at it :)